File: idatt2104-nettverksprogrammering/Nettverksprogrammering/examples/04-coroutines/await4.cpp
// Awaitable and await

#include <asio.hpp>
#include <iostream>

using namespace std;

asio::io_context event_loop(1); // Create event_loop optimized for running in one thread

asio::awaitable<int> get_answer() {
  // Simulate heavy work
  asio::steady_timer timer(event_loop, 5s);
  co_await timer.async_wait(asio::use_awaitable);
  co_return 42;
}

/// Using co_await to get a result asynchronously
asio::awaitable<void> event_loop_main() {
  auto answer = co_await get_answer();
  cout << answer << endl;
}

int main() {
  co_spawn(event_loop, event_loop_main, asio::detached);

  event_loop.run(); // Run event_loop in main thread
}

File: idatt2104-nettverksprogrammering/Nettverksprogrammering/examples/04-coroutines/await4.js
// Async and await

function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}

async function getAnswer() {
  await sleep(5000); // Simulate heavy work
  return 42;
}

let answer = await getAnswer();
console.log(answer);
